planning/louketo-proxy-plans.md

```diff
@@ -0,0 +1,420 @@
+
+    * Callback - callback endpoints for login and logout
+
+    * Health - health endpoint, including option to configure a health endpoint for the application
+
+    * Metrics - Prometheus metrics endpoint providing HTTP request statistics
+
+* CSRF protection
+
+## Service
+
+* Secure incoming request with bearer token, see token processing for details
+
+* Secure incoming request with session when service is deployed on the same domain as the application
+
+* Support external PDP
+
+* Support setting CORS headers
+
+## Token processing
+
+* Key management
+
+    * Obtain public keys from IdP (jwks_url)
+
+        * At startup
+
+        * Refresh keys automatically on demand (unknown kid)
+
+    * Supply signing keys directly in config
+
+* Support for non-opaque tokens (JWS and JWE)
+
+    * Should not require a request to IdP per incoming request
+
+    * Validate token using keys from IdP
+
+    * Support standard signing and encryption algorithms as specified in JWK specification
+
+* Support for opaque tokens
+
+    * Validate token using token introspection endpoint provided by IdP
+
+    * Support caching token introspection responses to reduce requests to IdP
+
+        * Locally
+
+        * External cache such as Infinispan or Reddis
+
+* Support validating/requiring standard and custom claims based on request path
+
+## Outgoing Web Proxy
+
+* Provide an outgoing web proxy to add security context (bearer token) to outgoing request
+
+* Add bearer token to outgoing request, where bearer token is obtained from one of the following
+
+    * Corresponding incoming request
+
+    * Service account (client credential grant)
+
+    * Secure Token Service
+
+## Kubernetes Operator
+
+* Automatically secure applications through a Kubernetes Operator
+
+    * Application defines a custom CR, or annotates the deployment
+
+    * Operator injects the proxy as a sidecar and updates service ports
+
+* Automatically register clients in IdP through a custom CR or using dynamic client registration
+
+## Plugins/extensibility
+
+* Plugin custom store for sessions and tokens
+
+* Define custom security headers
+
+* Define custom headers sent to upstream
+
+* Custom claim processing
+
+* Custom request/response handlers
+
+# Relevant specifications
+
+* OpenID Connect Core
+
+* OpenID Connect Discovery
+
+* OpenID Connect Session Management (limited to RP initiated logout)
+
+* OpenID Connect Front-Channel Logout
+
+* OpenID Connect Back-Channel Logout
+
+* OpenID Connect Dynamic Registration
+
+* JSON Web Signature (JWS)
+
+* JSON Web Encryption (JWE)
+
+* JSON Web Key (JWK)
+
+* OAuth 2.0 Token Introspection
+
+* OAuth 2.0 Security Best Current Practice
+
+* Financial-grade API - Part 1: Read-Only API Security Profile
+
+* Financial-grade API - Part 2: Read and Write API Security Profile
+
+* User Managed Access 2.0
+
+* OAuth 2.0 Mutual TLS Client Authentication and Certificate Bound Access Tokens
+
+# Review options
```

Comment:
@stianst I would like to remove this whole section related with `Envoy`. Wdyt?

Reply:
+1